#### COMPANION WORKBOOK

# Pandas

To make the most out this program, we strongly recommend you to:
1. First practice writing and implementing all of the code from Coding Section of the online lesson.
2. Then, freely experiment with and explore any interesting or confusing concepts. Simply insert new code cells and then use the help of Google and official documentation.
3. Finally, tackle all of the exercises at the end. They will help you tie everything together and **learn in context.**

#### <span style="color:#555">LESSON CODE SANDBOX</span>

Use this space to practice writing and implementing all of the code from Coding Section of the online lesson. Insert new code cells as needed, and feel free to write notes to yourself in Markdown.

## I. Pandas DataFrames are like spreadsheets.

In [4]:
import pandas as pd

In [5]:
example_dataframe = pd.DataFrame({
        'column_1' : [5, 4, 3],
        'column_2' : ['a', 'b', 'c']
    })

print(example_dataframe)

   column_1 column_2
0         5        a
1         4        b
2         3        c


In [6]:
# Read the iris dataset from a CSV file
df = pd.read_csv('project_files/iris.csv')

# Print data type for df
print( type(df) )

<class 'pandas.core.frame.DataFrame'>


In [7]:
df.head(10)
df.shape
df.sum(numeric_only=True)
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

## II. Pandas Series are like single columns.

In [3]:
integer_series = pd.Series([1,2,3,5])
print(integer_series)
print(integer_series.describe())

0    1
1    2
2    3
3    5
dtype: int64
count    4.000000
mean     2.750000
std      1.707825
min      1.000000
25%      1.750000
50%      2.500000
75%      3.500000
max      5.000000
dtype: float64


In [9]:
sepal_length = df["sepal_length"]
print(sepal_length)
print(type(sepal_length))

0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal_length, Length: 150, dtype: float64
<class 'pandas.core.series.Series'>


In [13]:
unique = sepal_length.unique()
print(unique)

petal_area = (pd.Series) (df["petal_length"] * df["petal_width"])
df["petal_area"] = petal_area
print(petal_area.describe())

[5.1 4.9 4.7 4.6 5.  5.4 4.4 4.8 4.3 5.8 5.7 5.2 5.5 4.5 5.3 7.  6.4 6.9
 6.5 6.3 6.6 5.9 6.  6.1 5.6 6.7 6.2 6.8 7.1 7.6 7.3 7.2 7.7 7.4 7.9]
count    150.000000
mean       5.794067
std        4.712390
min        0.110000
25%        0.420000
50%        5.615000
75%        9.690000
max       15.870000
dtype: float64


## III. Boolean masks allow you to filter data.

In [22]:
# Series
example_series = pd.Series([10, 5, -3, 2])
print(example_series)

# Mask where value > 0
series_mask = (pd.Series) (example_series > 0)
print(series_mask)

# Series where value > 0
print(example_series[series_mask])

# Series inverted
print(example_series[~series_mask])

0    10
1     5
2    -3
3     2
dtype: int64
0     True
1     True
2    False
3     True
dtype: bool
0    False
1    False
2     True
3    False
dtype: bool
0    10
1     5
3     2
dtype: int64
2   -3
dtype: int64


In [30]:
# Display observations where petal_area > 14 and petal_length >= 6
df[(df.petal_area > 14) & (df.petal_length >= 6)]

sepal_length  sepal_width  petal_length  petal_width    species  \
100           6.3          3.3           6.0          2.5  virginica   
109           7.2          3.6           6.1          2.5  virginica   
117           7.7          3.8           6.7          2.2  virginica   
118           7.7          2.6           6.9          2.3  virginica   
135           7.7          3.0           6.1          2.3  virginica   

     petal_area  
100       15.00  
109       15.25  
117       14.74  
118       15.87  
135       14.03

In [45]:
example_mask = pd.Series([True,True,False,True,False])
example_indicators = example_mask.astype(int)
print(example_indicators)

df["giant"] = (df.petal_area > 14).astype(int)
df[df.giant == 1].describe()

# Versicolor or virginica
species_mask = df.species.isin(['versicolor', 'virginica'])

# Sepal width > 3.2
sepal_width_mask = df.sepal_width > 3.2

# Index with both masks
df[species_mask & sepal_width_mask]

0    1
1    1
2    0
3    1
4    0
dtype: int32


sepal_length  sepal_width  petal_length  petal_width     species  \
56            6.3          3.3           4.7          1.6  versicolor   
85            6.0          3.4           4.5          1.6  versicolor   
100           6.3          3.3           6.0          2.5   virginica   
109           7.2          3.6           6.1          2.5   virginica   
117           7.7          3.8           6.7          2.2   virginica   
124           6.7          3.3           5.7          2.1   virginica   
131           7.9          3.8           6.4          2.0   virginica   
136           6.3          3.4           5.6          2.4   virginica   
144           6.7          3.3           5.7          2.5   virginica   
148           6.2          3.4           5.4          2.3   virginica   

     petal_area  giant  
56         7.52      0  
85         7.20      0  
100       15.00      1  
109       15.25      1  
117       14.74      1  
124       11.97      0  
131       12.80      0  
136       13.44      0  
144       14.25      1  
148       12.42      0

array(['setosa', 'versicolor', 'virginica'], dtype=object)

In [119]:
df_group = df.groupby('species').agg(['min', 'median', 'mean', 'std', 'max', 'skew']).petal_area

petal_area_kurt = []
for idx, species in enumerate(df.species.unique()):
    petal_area_kurt.append(df[df.species == species].petal_area.kurt())

df_group["kurt"] = petal_area_kurt
df_group

min  median     mean       std    max      skew      kurt
species                                                               
setosa      0.11   0.300   0.3656  0.181155   0.96  1.440357  2.071190
versicolor  3.30   5.615   5.7204  1.368403   8.64 -0.012948 -0.510671
virginica   7.50  11.445  11.2962  2.157412  15.87  0.130633 -0.783504

sepal_length    -0.552064
sepal_width      0.228249
petal_length    -1.402103
petal_width     -1.340604
petal_area      -1.148879
giant           20.767212
dtype: float64

## IV. Groupbys allow you to segment and aggregate data.

#### <span style="color:#555">EXERCISES</span>

Complete each of the following exercises.

First, remember to import Pandas and read in the iris dataset.

In [ ]:
import pandas as pd

df = pd.read_csv('project_files/iris.csv')

## <span style="color:RoyalBlue">Exercise 5.1 - Flowers for Casey</span>

In the previous lesson, we discovered that the first local business we need to visit in Park Royal is a local **flower shop.** When we arrive, we find out that the owner, Casey, urgently needs our help!

They just received a new shipment of **iris flowers**, but they've never stocked these flowers before. Casey asks us to share what we know about these new flowers. Luckily, we have the *iris dataset* to learn more about them ourselves!

In the previous lesson, we also introduced the idea of using a **toy problem** to reduce the size of a problem and break down the concepts. While 150 observations is not exactly "big data," it's still too large to fit on our screen at once. So let's use another toy problem to practice the concepts in this lesson.

#### A.) First, create a new DataFrame called <code style="color:steelblue">toy_df</code>. It should contain the first 5 rows plus the last 5 rows from our original Iris dataset.
* **Tip:** You already have a <code style="color:steelblue">.head()</code>, but what about a <code style="color:steelblue">.tail()</code>?
* **Tip:** <code style="color:steelblue">pd.concat()</code> is your [friend](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html).

In [125]:
# Create toy_df
toy_df = pd.DataFrame(pd.concat([df.head(5),df.tail(5)]))

#### B.) Next, display <code style="color:steelblue">toy_df</code>.
* After all, it will only be 10 rows.
* In <code style="color:steelblue">toy_df</code>, you should have data from 2 different species of flower. Which are they?

In [124]:
# Display toy_df
toy_df

sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  giant  
0          0.28      0  
1          0.28      0  
2          0.26      0  
3          0.30      0  
4          0.28      0  
145       11.96      0  
146        9.50      0  
147       10.40      0  
148       12.42      0  
149        9.18      0

<strong style="color:RoyalBlue">Expected output:</strong>

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
    </tr>
  </tbody>
</table>

#### C.) Next, display a summary table for <code style="color:steelblue">toy_df</code>.
* It should have the mean, standard deviation, and quartiles for each of the columns.
* Since <code style="color:steelblue">toy_df</code> is only 10 rows, you can manually check the **mins** and **maxes**. Are they correct?

In [128]:
# Describe toy_df
toy_df.agg(['min','max'])

sepal_length  sepal_width  petal_length  petal_width    species  \
min           4.6          2.5           1.3          0.2     setosa   
max           6.7          3.6           5.4          2.3  virginica   

     petal_area  giant  
min        0.26      0  
max       12.42      0

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>10.000000</td>
      <td>10.000000</td>
      <td>10.000000</td>
      <td>10.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>5.590000</td>
      <td>3.130000</td>
      <td>3.290000</td>
      <td>1.130000</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.807534</td>
      <td>0.316403</td>
      <td>1.995244</td>
      <td>0.992248</td>
    </tr>
    <tr>
      <th>min</th>
      <td>4.600000</td>
      <td>2.500000</td>
      <td>1.300000</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>4.925000</td>
      <td>3.000000</td>
      <td>1.400000</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>5.500000</td>
      <td>3.050000</td>
      <td>3.250000</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>6.275000</td>
      <td>3.350000</td>
      <td>5.175000</td>
      <td>1.975000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>6.700000</td>
      <td>3.600000</td>
      <td>5.400000</td>
      <td>2.300000</td>
    </tr>
  </tbody>
</table>

## <span style="color:RoyalBlue">Exercise 5.2 - Basic Feature Engineering</span>

Elementwise operations are very useful in machine learning, especially for **feature engineering.** Feature engineering is the process of creating new features (model input variables) from existing ones, and its one of the best ways for data scientists to add value.

In the Iris dataset, we have petal width and length, but what if we wanted to know petal area? Well, we can create a new <code style="color:steelblue">petal_area</code> feature (yes, the petals are not perfect rectangles, but that's fine).

#### A.) First, display the two columns of <code style="color:steelblue">petal_width</code> and <code style="color:steelblue">petal_length</code> in <code style="color:steelblue">toy_df</code>.
* **Tip:** You can index a DataFrame using a list of column names too, like so:


<pre style="color:steelblue">df[['column_1', 'column_2']]</pre>

In [129]:
# Display petal_width and petal_length
toy_df[['petal_width','petal_length']]

petal_width  petal_length
0            0.2           1.4
1            0.2           1.4
2            0.2           1.3
3            0.2           1.5
4            0.2           1.4
145          2.3           5.2
146          1.9           5.0
147          2.0           5.2
148          2.3           5.4
149          1.8           5.1

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>petal_width</th>
      <th>petal_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.2</td>
      <td>1.3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.2</td>
      <td>1.5</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>145</th>
      <td>2.3</td>
      <td>5.2</td>
    </tr>
    <tr>
      <th>146</th>
      <td>1.9</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>147</th>
      <td>2.0</td>
      <td>5.2</td>
    </tr>
    <tr>
      <th>148</th>
      <td>2.3</td>
      <td>5.4</td>
    </tr>
    <tr>
      <th>149</th>
      <td>1.8</td>
      <td>5.1</td>
    </tr>
  </tbody>
</table>

#### B.) Next, create a new <code style="color:steelblue">petal_area</code> feature in <code style="color:steelblue">toy_df</code>.
* Multiply the <code style="color:steelblue">petal_width</code> column by the <code style="color:steelblue">petal_length</code> column.
* Display <code style="color:steelblue">toy_df</code> after creating the new feature.
* Are the values for <code style="color:steelblue">petal_area</code> correct? Manually spot check a few of them just to make sure.

In [130]:
# Create a new petal_area column
toy_petal_area = toy_df.petal_length * toy_df.petal_width
toy_df["petal_area"] = toy_petal_area

# Display toy_df
toy_df


sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  giant  
0          0.28      0  
1          0.28      0  
2          0.26      0  
3          0.30      0  
4          0.28      0  
145       11.96      0  
146        9.50      0  
147       10.40      0  
148       12.42      0  
149        9.18      0

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.26</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.30</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>11.96</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>9.50</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>10.40</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>12.42</td>
    </tr>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
      <td>9.18</td>
    </tr>
  </tbody>
</table>

### C.) Finally, what do we now know about Iris flowers?

By creating a <code style="color:steelblue">petal_area</code> feature, it's now much easier to see that virginica flowers have significantly larger petals than setosa flowers do! Often, by creating new features, you can learn more about the data (and improve your machine learning models). As you might guess, a machine learning algorithm and take these features and tease out the underlying patterns in a much more precise and comprehensive way.

## <span style="color:RoyalBlue">Exercise 5.3 - Mask On Mask Off</span>

Now, let's use the <code style="color:steelblue">toy_df</code> to really drive home the concept of **boolean masks**, because we'll be using these very frequently in data science.

For example, assume we wanted to display observations where <code style="color:steelblue">petal_area > 10</code> and <code style="color:steelblue">sepal_width > 3</code>. How could we do so?

#### A.) First, display <code style="color:steelblue">toy_df</code> again just to have it in front of you.

In [131]:
# Display toy_df
toy_df

sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  giant  
0          0.28      0  
1          0.28      0  
2          0.26      0  
3          0.30      0  
4          0.28      0  
145       11.96      0  
146        9.50      0  
147       10.40      0  
148       12.42      0  
149        9.18      0

#### B.) Take a look at the DataFrame and manually count the number that satisfy our conditions.
* How many observations have <code style="color:steelblue">petal_area > 10</code>?
* How many observations have <code style="color:steelblue">sepal_width > 3</code>?
* How many satisfy both conditions?

*In the toy_df, there are:*
* **3** observations with <code style="color:steelblue">petal_area > 10</code> (indices 145, 147, 148)
* **5** observations with <code style="color:steelblue">sepal_width > 3</code> (indices 0,2,3,4,148)
* **1** observation that satifies both conditions (index 148)

Great. Now we'll see what's going on under the hood when we use our boolean masks.

#### C.) Create a boolean mask for <code style="color:steelblue">petal_area > 10</code>.
* Name it <code style="color:steelblue">petal_area_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [133]:
# Mask for petal_area > 10
toy_petal_area_mask = toy_df.petal_area > 10

# Display petal_area_mask
toy_petal_area_mask

0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool
</pre>

#### D.) Next, create a boolean mask for <code style="color:steelblue">sepal_width > 3</code>.
* Name it <code style="color:steelblue">sepal_width_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [134]:
# Mask for sepal_width > 3
toy_sepal_width_mask = toy_df.sepal_width > 3

# Display sepal_width_mask
toy_sepal_width_mask

0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool
</pre>

#### E.) Next, display the two masks combined using the <code style="color:steelblue">&</code> operator.
* Note how their combination results in another boolean mask!

In [137]:
# Display both masks, combined
toy_mask = toy_petal_area_mask & toy_sepal_width_mask
toy_mask

0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool
</pre>

#### F.) Finally, select the observation(s) from <code style="color:steelblue">toy_df</code> where both conditions are met.
* Do you get the result you expected based on (B)?

In [138]:
# Index with both masks
toy_df[toy_mask]

sepal_length  sepal_width  petal_length  petal_width    species  \
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  giant  
148       12.42      0

## <span style="color:RoyalBlue">Exercise 5.4 - Iris Insights</span>

With the help of **groupbys**, we're almost ready to return to Casey and share what we've learned about Iris flowers. But before we do, let's just bring back our <code style="color:steelblue">toy_df</code> for one last hoorah, just to make sure we know what's going on under the hood.

Let's calculate the median <code style="color:steelblue">petal_area</code> for each species. Since <code style="color:steelblue">toy_df</code> is small, we can do this manually as well and check to make sure the values are correct.


#### A.) First, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the virginica flowers in our <code style="steelblue">toy_df</code>.
* Display all observations of the virginica species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* **Tip:** Check out the <code style="color:steelblue">.sort_values()</code> function ([documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)).

In [140]:
# Display all 'virginica' species, sorted by petal_area
toy_df[toy_df.species == 'virginica']

sepal_length  sepal_width  petal_length  petal_width    species  \
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  giant  
145       11.96      0  
146        9.50      0  
147       10.40      0  
148       12.42      0  
149        9.18      0

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
      <td>9.18</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>9.50</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>10.40</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>11.96</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>12.42</td>
    </tr>
  </tbody>
</table>

Based on the output above, what's median <code style="color:steelblue">petal_area</code> for the virginica species?

#### B.) Next, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the setosa flowers in our <code style="steelblue">toy_df</code>.
* Display all observations of the setosa species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* Based on the output, what's median <code style="color:steelblue">petal_area</code> for the setosa species?

In [141]:
# Display all 'setosa' species
toy_df[toy_df.species == 'setosa']

sepal_length  sepal_width  petal_length  petal_width species  petal_area  \
0           5.1          3.5           1.4          0.2  setosa        0.28   
1           4.9          3.0           1.4          0.2  setosa        0.28   
2           4.7          3.2           1.3          0.2  setosa        0.26   
3           4.6          3.1           1.5          0.2  setosa        0.30   
4           5.0          3.6           1.4          0.2  setosa        0.28   

   giant  
0      0  
1      0  
2      0  
3      0  
4      0

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.26</td>
    </tr>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.30</td>
    </tr>
  </tbody>
</table>

#### C.) Finally let's calculate the median values using a <code style="color:steelblue">.groupby()</code>.
* Do you get the same result?

In [144]:
# Median petal_area in toy_df
toy_df.groupby('species').median().petal_area

species
setosa        0.28
virginica    10.40
Name: petal_area, dtype: float64

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
species
setosa        0.28
virginica    10.40
Name: petal_area, dtype: float64
</pre>

#### Congratulations... You've completed the training mission!

In this lesson, you explored the Iris dataset using Pandas. After arming yourself with Pandas, you've successfully completed the rest of the training mission!*